File: sdr_receiver_py_wrapper/sdr_receiver_py_wrapper/mock_radar_context_publisher.py
from __future__ import annotations

import rclpy
from rclpy.node import Node
from rclpy.qos import HistoryPolicy, QoSProfile, ReliabilityPolicy

from sdr_receiver.msg import JamCode, RadarContext


def level_to_radar_info(level: int, *, key_mutable: bool = True) -> int:
    raw = (int(level) & 0x03) << 3
    if key_mutable:
        raw |= 1 << 5
    return raw & 0xFF


class MockRadarContextPublisher(Node):
    def __init__(self) -> None:
        super().__init__("mock_radar_context_publisher_py")
        self.declare_parameter("self_id", 9)
        self.declare_parameter("start_level", 1)
        self.declare_parameter("max_level", 3)
        self.declare_parameter("publish_hz", 5.0)
        self.declare_parameter("auto_advance_on_jam_code", True)
        self.declare_parameter("topic", "/judge/radar_context")

        self.self_id = int(self.get_parameter("self_id").value)
        self.level = int(self.get_parameter("start_level").value)
        self.max_level = int(self.get_parameter("max_level").value)
        self.auto_advance = bool(self.get_parameter("auto_advance_on_jam_code").value)
        topic = str(self.get_parameter("topic").value)

        qos = QoSProfile(
            history=HistoryPolicy.KEEP_LAST,
            depth=5,
            reliability=ReliabilityPolicy.RELIABLE,
        )
        self.pub = self.create_publisher(RadarContext, topic, qos)
        self.sub = self.create_subscription(JamCode, "/sdr/jam_code", self._on_jam_code, qos)
        period = 1.0 / max(0.1, float(self.get_parameter("publish_hz").value))
        self.timer = self.create_timer(period, self._publish_context)
        self.get_logger().info(
            f"mock RadarContext publishing on {topic}, self_id={self.self_id}, level={self.level}"
        )

    def _publish_context(self) -> None:
        msg = RadarContext()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.self_id = self.self_id
        msg.self_color = 2 if self.self_id < 100 else 0
        msg.radar_info_raw = level_to_radar_info(self.level)
        msg.jam_level = self.level
        msg.key_mutable = True
        msg.game_progress = 4
        msg.match_time = 420
        msg.referee_online = True
        self.pub.publish(msg)

    def _on_jam_code(self, msg: JamCode) -> None:
        if not self.auto_advance or not msg.valid:
            return
        if int(msg.level) == self.level and self.level < self.max_level:
            self.level += 1
            self.get_logger().info(f"mock context advanced to L{self.level}")


def main(args=None) -> None:
    rclpy.init(args=args)
    node = MockRadarContextPublisher()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: sdr_receiver/src/mock_radar_context_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <sdr_receiver/msg/jam_code.hpp>
#include <sdr_receiver/msg/radar_context.hpp>

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <memory>
#include <string>

class MockRadarContextPublisher final : public rclcpp::Node {
public:
    MockRadarContextPublisher() : Node("mock_radar_context_publisher") {
        topic_ = declare_parameter<std::string>("topic", "/judge/radar_context");
        jam_code_topic_ = declare_parameter<std::string>("jam_code_topic", "/sdr/jam_code");
        self_id_ = static_cast<uint8_t>(declare_parameter<int>("self_id", 9));
        const int start_level = static_cast<int>(declare_parameter<int>("start_level", 1));
        const int max_level = static_cast<int>(declare_parameter<int>("max_level", 3));
        level_ = std::clamp(start_level, 0, 3);
        max_level_ = std::clamp(max_level, 1, 3);
        key_mutable_ = declare_parameter<bool>("key_mutable", true);
        referee_online_ = declare_parameter<bool>("referee_online", true);
        advance_on_jam_code_ = declare_parameter<bool>("advance_on_jam_code", true);
        const int period_param_ms = static_cast<int>(declare_parameter<int>("period_ms", 200));
        const int period_ms = std::max(20, period_param_ms);

        pub_ = create_publisher<sdr_receiver::msg::RadarContext>(topic_, rclcpp::QoS(rclcpp::KeepLast(5)).reliable());
        sub_ = create_subscription<sdr_receiver::msg::JamCode>(
            jam_code_topic_,
            10,
            [this](const sdr_receiver::msg::JamCode::SharedPtr msg) {
                if (!advance_on_jam_code_ || !msg->valid) return;
                if (msg->level == level_ && level_ < max_level_) {
                    ++level_;
                    RCLCPP_INFO(get_logger(), "received L%d key, advancing mock context to L%d", msg->level, level_);
                } else if (msg->level >= max_level_) {
                    RCLCPP_INFO(get_logger(), "received final L%d key; holding mock context at max level", msg->level);
                }
            });

        timer_ = create_wall_timer(std::chrono::milliseconds(period_ms), [this]() { publish_context(); });
        RCLCPP_INFO(get_logger(), "publishing mock RadarContext on %s, self_id=%u, start_level=%d",
                    topic_.c_str(), static_cast<unsigned>(self_id_), level_);
    }

private:
    void publish_context() {
        sdr_receiver::msg::RadarContext msg;
        msg.header.stamp = now();
        msg.header.frame_id = "mock_referee";
        msg.self_id = self_id_;
        msg.self_color = self_id_ >= 100 ? 0 : 2;
        msg.radar_info_raw = static_cast<uint8_t>((level_ & 0x03) << 3);
        if (key_mutable_) msg.radar_info_raw |= static_cast<uint8_t>(1u << 5);
        msg.jam_level = static_cast<uint8_t>(level_);
        msg.key_mutable = key_mutable_;
        msg.game_progress = 4;
        msg.match_time = 300;
        msg.referee_online = referee_online_;
        pub_->publish(msg);
    }

    rclcpp::Publisher<sdr_receiver::msg::RadarContext>::SharedPtr pub_;
    rclcpp::Subscription<sdr_receiver::msg::JamCode>::SharedPtr sub_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::string topic_;
    std::string jam_code_topic_;
    uint8_t self_id_ = 9;
    int level_ = 1;
    int max_level_ = 3;
    bool key_mutable_ = true;
    bool referee_online_ = true;
    bool advance_on_jam_code_ = true;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MockRadarContextPublisher>());
    rclcpp::shutdown();
    return 0;
}
